www/index.js

```diff
@@ -1,86 +1,171 @@
+      return Math.hypot(
+        rectB.left - rectA.right,
+        (centerB.y - centerA.y) * verticalOrthogonalWeight
+      );
   }
 }
 
-class ChannelsListElement extends ReactiveElement {
-  static get reactiveProperties() {
-    return {
-      channels: Object,
-    };
+/**
+ * @param {DOMRect} targetRect
+ * @param {DOMRect} sourceRect
+ * @param {string} direction
+ */
+function isValidForDirection(targetRect, sourceRect, direction) {
+  switch (direction) {
+    case 'up':
+      return targetRect.bottom < sourceRect.top;
+    case 'down':
+      return targetRect.top > sourceRect.bottom;
+    case 'left':
+      return targetRect.right < sourceRect.left;
+    case 'right':
+      return targetRect.left > sourceRect.right;
   }
+}
 
-  async updated() {
-    render(this.render(), this);
+/**
+ * @param {HTMLElement} source
+ * @param {Iterable<HTMLElement>} candidates
+ * @param {string} direction
+ * @returns {HTMLElement}
+ */
+function getClosestInDirection(source, candidates, direction) {
+  const sourceRect = source.getBoundingClientRect();
 
-    await Promise.resolve();
+  let closest;
+  let minDistance = Number.POSITIVE_INFINITY;
 
-    console.log(
-      `ChannelsListElement .channel count: ${
-        this.querySelectorAll('.channel').length
-      }`
-    );
-  }
+  for (const candidate of candidates) {
+    if (candidate !== source) {
+      const candidateRect = candidate.getBoundingClientRect();
+
+      if (isValidForDirection(candidateRect, sourceRect, direction)) {
+        const distance = getDistanceInDirection(
+          candidateRect,
+          sourceRect,
+          direction
+        );
 
-  render() {
-    return html`<div>
-      <h2>Channel list</h2>
-      <tv-scroll-container>
-        ${this.channels.map(
-          (channel) => html`<tv-channel .channel="${channel}"></tv-channel>`
-        )}
-      </tv-scroll-container>
-    </div>`;
+        console.debug(
+          'Distance from',
+          source.title,
+          'to',
+          candidate.title,
+          'is',
+          distance
+        );
+
+        if (distance < minDistance) {
+          minDistance = distance;
+          closest = candidate;
+        }
+      }
+    }
   }
-}
 
-class LivePageElement extends ReactiveElement {
-  async updated() {
-    render(this.render(), this);
+  return closest;
+}
 
-    await Promise.resolve();
+function handleKeyDown(event) {
+  const direction = directionByKey[event.key];
 
-    console.log(
-      `LivePageElement .channel count: ${
-        this.querySelectorAll('.channel').length
-      }`
-    );
+  if (!direction) {
+    return;
   }
 
-  render() {
-    const channels = [
-      { id: 'nrk1', name: 'NRK1' },
-      { id: 'nrk2', name: 'NRK2' },
-      { id: 'nrk3', name: 'NRK3' },
-    ];
-
-    return html`<div>
-      <h1>Live page</h1>
-      <tv-channels-list .channels=${channels}></tv-channels-list>
-    </div>`;
+  document.querySelectorAll('[data-focused]').forEach((
```

Comment:
During this experimentation phase, it might be interesting to debug how long this searching operation takes